scipy/optimize/optimize.py

```diff
@@ -287,6 +287,24 @@ def _check_clip_x(x, bounds):
     return x
 
 
+def _filter_params_func(func, full_x, mask):
+    # Wraps a function expecting a full set of x0 into a function that
+    # accepts a reduced parameter set, based on a mask
+    def wrapped_func(x, *args):
+        expanded_x = np.copy(full_x)
+        expanded_x[~mask] = x
+        return func(expanded_x, *args)
+
+    return wrapped_func
+
+
+def _filter_params_jac(jac, mask):
```

Comment:
Even though the parameter that specifies the gradient function is called `jac`, I think we should use the term `grad` for the partials of a scalar objective function and reserve `jac` for the partials of a vector-valued function.

On a related note, this doesn't yet work for the jacobian of vector-valued constraints.
```
import numpy as np
import scipy.optimize as optimize

def f(x):
    return (x**3).sum()

def con(x):
    return x**2-1

def conjac(x):
    return np.diag(2*x)

def callback(x):
    print(x.shape)

method = 'slsqp'
constraints={"fun":con, "type":'ineq', "jac":conjac}
x0 = np.array([0.5, 3.0, 2, 5])
bounds = [(0.0, 3.0), (2.0, 2), (0, np.inf), (1.5, 1.5)]

res = optimize.minimize(
    f, x0, method=method, bounds=bounds, constraints=constraints, callback=callback
)
```

Reply:
SLSQP constraints should be scalar functions, `R^n --> 1`. You can have more than one constraint, but each one needs to be scalar.

Reply:
jac and grad are littered around the codebase all over the place, I get your point.
Fixed the vector valued constraint problem and added a test.